File: ListaEncadeada/ListaEncadeada.cs
namespace ListaEncadeada;
public class Lista
{
  private No head;
  private int qtd;

  public Lista()
  {
    head = new No("a");
    qtd = 0;
  }
  public No Head() => head;
  public int Qtd() => qtd;
   public void inserirNoInicio(No n)
  {
    n.setNext(head.getNext());
    head.setNext(n);
    qtd++;
  }

  public void inserirNoFinal(No n)
  {
    No aux = head;
    while (aux.getNext() != null) aux = aux.getNext();
    aux.setNext(n);
    qtd++;
  }
  // la ele
  public void inserirNaPos(int pos, No n)
  {
    if (pos < 0)
    {
      return;
    }
    if (pos > qtd) pos = qtd - 1;
    if (pos == 0)
    {
      n.setNext(head);
      head = n;
      qtd++;
      return;
    }
    No aux = head;
    for (int i = 0; i < pos - 1; pos++)
    {
      aux = aux.getNext();
    }
    n.setNext(aux.getNext());
    aux.setNext(n);
    qtd++;
  }

  public void inserirOrdenado(No n)
  {
    if (string.Compare(n.getElement(), head.getElement()) == -1)
    {
      n.setNext(head);
      head = n;
      qtd++;
      return;
    }
    No anterior = head;
    No aux = head.getNext();
    while (string.Compare(n.getElement(), aux.getElement()) == 1)
    {
      anterior = aux;
      aux = aux.getNext();
    }
    n.setNext(aux);
    anterior.setNext(n);
    qtd++;
  }

  public void deletarNoInicio() {
    if(head.getNext() == null) return;
    head = head.getNext();
    qtd--;
  }

  public void deletarNoFinal() {
    No aux = head;
    while(aux.getNext().getNext() != null) {
      aux = aux.getNext();
    }
    aux.setNext(null);
    qtd--;
  }

  public void deletarNaPos(int pos) {
    if(pos  < 0) return;
    if(pos > qtd) pos = qtd -1;
    if(pos == 0) head = head.getNext();
    No anterior = head;
    No aux = head.getNext();
    for(int i = 0; i  < pos -1; i++ ) {
      anterior = aux;
      aux = aux.getNext();
    } 
    anterior.setNext(aux.getNext());
    qtd--;
  }
  public No Buscar(int pos) {
    if(pos < 0) return null;
    if(pos > qtd) pos = qtd -1;
    No aux = head;
    for(int i = 0; i < pos; i++) {
      aux = aux.getNext();
    }
    return aux;
  }
  public void printar()
  {
    No aux = head;
    int i = 0;
    while (aux != null)
    {
      Console.WriteLine($"{i} -> {aux.getElement()}");
      aux = aux.getNext();
      i++;
    }
  }
}

File: ListaEncadeada/lista_encadeada.go
package listaencadeada

import (
	"fmt"
	"strings"
)

type No struct {
	element string
	next    *No
}

func (n *No) Next() *No {
	return n.next
}
func (n *No) Element() string { return n.element }

type ListaEncadeada struct {
	head *No
	qtd  int
}

func (l *ListaEncadeada) Head() *No {
	return l.head
}

func NewListaEncadeada() *ListaEncadeada {
	return &ListaEncadeada{head: &No{element: "a", next: nil}}
}

func (l *ListaEncadeada) InserirNoInicio(element string) {
	no := &No{element: element}
	no.next = l.head.next
	l.head.next = no
	l.qtd++
}

func (l *ListaEncadeada) InserirNoFinal(element string) {
	aux := l.head
	for aux.next != nil {
		aux = aux.next
	}
	aux.next = &No{element: element}
	l.qtd++
}

// la ele
func (l *ListaEncadeada) InserirNaPos(pos int, element string) {
	n := &No{element: element}
	if pos < 0 {
		return
	}
	if pos > l.qtd {
		pos = l.qtd - 1
	}
	if pos == 0 {
		n.next = l.head
		l.head = n
		l.qtd++
		return
	}
	count := 1
	aux := l.head
	for count < pos {
		aux = aux.next
		count++
	}
	n.next = aux.next
	aux.next = n
	l.qtd++
}
func (l *ListaEncadeada) InserirOrdenado(element string) {
	n := &No{element: element}
	if strings.Compare(n.element, l.head.element) == -1 {
		n.next = l.head
		l.head = n
		l.qtd++
		return
	}
	anterior := l.head
	aux := l.head.next
	for strings.Compare(n.element, aux.element) == 1 {
		anterior = aux
		aux = aux.next
	}
	n.next = aux
	anterior.next = n
	l.qtd++
}

func (l *ListaEncadeada) DeletarNoInicio() {
	if l.head.next == nil {
		return
	}
	l.head = l.head.next
	l.qtd--
}

func (l *ListaEncadeada) DeletarNoFinal() {
	aux := l.head
	for aux.next.next != nil {
		aux = aux.next
	}
	aux.next = nil
	l.qtd--
}

func (l *ListaEncadeada) DeletarNaPos(pos int) {
	if pos < 0 {
		return
	}
	if pos > l.qtd {
		pos = l.qtd - 1
	}
	if pos == 0 {
		l.head = l.head.next
		l.qtd--
		return
	}
	count := 1
	anterior := l.head
	aux := l.head.next
	for count < pos {
		anterior = aux
		aux = aux.next
		count++
	}
	anterior.next = aux.next
	l.qtd--
}

func (l *ListaEncadeada) Buscar(pos int) *No {
	if pos < 0 {
		return nil
	}
	if pos > l.qtd {
		pos = l.qtd - 1
	}
	count := 0
	aux := l.head
	for count < pos {
		aux = aux.next
		count++
	}
	return aux
}

func (l *ListaEncadeada) Printar() {
	i := 0
	aux := l.head
	for aux != nil {
		fmt.Printf("%d -> %s\n", i, aux.element)
		aux = aux.next
		i++
	}
}
